### ISSUE TITLE
Module resolution error on Windows

### ISSUE THREAD

**Issue by reinei** [react-cosmos/react-cosmos#288]
### What's up?

I am trying to load proxies in my (new and clean) project.
However, the webpack packing process fails with the following error:
```
ERROR in ./~/react-cosmos-webpack/lib/user-modules.js
edux-proxy.js' in 'D:\Development\centrum-sys\node_modules\react-cosmos-webpack\lib'
 @ ./~/react-cosmos-webpack/lib/user-modules.js 21:14-66
 @ ./~/react-cosmos-webpack/lib/entry.js
 @ multi (webpack)-hot-middleware/client.js?reload=true ./~/react-cosmos-webpack/lib/entry.js
```
(the error is literally stripped, but trying to load cosmos with only one proxy results in the full line being shown:
```Module not found: Error: Can't resolve 'D:Developmentcentrum-syscontext-proxy.js' in ```
note the missing slashes)

### Mkay, tell me more...

I am using the following libraries

* react-cosmos-context-proxy: 2.0.0-beta.6
* react-cosmos-redux-proxy: 2.0.0-beta.6
* react-cosmos-webpack: 2.0.0-beta.7
* webpack: 2.2.1
* webpack-dev-server: 2.3.0
* react: 15.4.2
* redux: 3.6.0

(all have been installed without version tags, which I assume means the beta builds are fine OR the last npm publish did not tag them correctly)

I am running node v7.2.0 on Win10 64bit.

my config files are attached below.
[config.zip](https://github.com/react-cosmos/react-cosmos/files/771136/config.zip)

I hope this is enough information to allow someone to point at my mistake

PS: is this even the right place to ask these questions?

**Comment by skidding** [comment]
Hi @reinei 

This is the right place. Unfortunately most of the users (myself included) are on OSX, so I apologize if the current version is broken on Windows. Maybe we can improve on this together while fixing your installation.

I'll start with some off the top of my head suggestions:
- What's weird about `'D:Developmentcentrum-syscontext-proxy.js'` is that you only have `./redux-proxy.js` in your `cosmos.config.js`, so where could `...context-proxy.js` come from?
- Can you try cloning react-cosmos and [running one of the examples](https://github.com/react-cosmos/react-cosmos/blob/master/CONTRIBUTING.md#get-familiar-with-codebase)? This should tell us whether we have a generic problem on Windows or if we should debug your config more.

Thanks!

**Comment by reinei** [comment]
Hey skidding,
1. The fact that the file only contains ./redux-proxy.js was my bad (didn't hit save in atom)
2. The ```npm i``` command in the main repository errors with the following log file
[npm-debug.txt](https://github.com/react-cosmos/react-cosmos/files/771229/npm-debug.txt)
It also says ```packages/$PACKAGE/src doesn't exist``` 4 times
(yes it spits out the error from the log 4 times as well as the above line)

Thanks in advance for helping

**Comment by skidding** [comment]
> The fact that the file only contains ./redux-proxy.js was my bad (didn't hit save in atom)

Just to be sure, do you have a `context-proxy.js` file in the same dir as `cosmos.config.js`?

**Comment by reinei** [comment]
Yes I have both proxies and sadly simply removing one of them does not solve my issue (both proxies are copies of the code from the readme.md)

**Comment by skidding** [comment]
> Yes I have both proxies and sadly simply removing one of them does not solve my issue (both proxies are copies of the code from the readme.md)

For the purpose of pinning the problem, does the build work if you remove them both (`proxies: []`)?

**Comment by reinei** [comment]
Yes (as stated) removing all proxies results in a successful build and execution in my own project

**Comment by skidding** [comment]
Can you please try resolving the proxy paths manually? E.g.

```js
// cosmos.config.js
proxies: [
  require.resolve('./context-proxy.js'),
  require.resolve('./redux-proxy.js'),
]
```

**Comment by reinei** [comment]
Manually resolving the paths does not change the error in any way (it still tries to load the proxies from
```node_modules\react-cosmos-webpack\lib```

**Comment by skidding** [comment]
> Manually resolving the paths does not change the error in any way (it still tries to load the proxies from node_modules\react-cosmos-webpack\lib

Weird. I'm curious how those paths are built. Seeing the contents of your resolved config should give us some insight:

![image](https://cloud.githubusercontent.com/assets/250750/22887441/c4adab2a-f20a-11e6-9d28-0c3e6210809b.png)

Try the same in your repo and paste the output here, but please note that I'm using a custom config path in my example `./cosmos/cosmos.config.js` so adjust it to match your structure.

**Comment by reinei** [comment]
My output is ```{ componentPaths: [ 'D:\\Development\\centrum-sys\\src\\components' ],
  fixturePaths: [],
  globalImports: [],
  hmrPlugin: true,
  hostname: 'localhost',
  hot: true,
  ignore: [],
  port: 8989,
  proxies:
   [ 'D:\\Development\\centrum-sys\\redux-proxy.js',
     'D:\\Development\\centrum-sys\\context-proxy.js' ],
  webpackConfigPath: 'D:\\Development\\centrum-sys\\webpack.config',
  publicPath: 'D:\\Development\\centrum-sys\\public' }```
Which is exactly what I expected for my windows (all slashes are escaped correctly ...)

**Comment by skidding** [comment]
OK, we're getting there...

The config looks good so the next step is the webpack loader which injects this paths into the bundle entry point. You can see the generated source for the entry point but opening `node_modules/react-cosmos-webpack/lib/user-modules.js`, going at the last line (105) and changing the code like this:

```diff
-return source.replace(/COMPONENTS/g, convertPathMapToRequireCalls(components)).replace(/FIXTURES/g, convertPathMapToRequireCalls(fixtures)).replace(/PROXIES/g, convertPathListToRequireCalls(proxies)).replace(/CONTEXTS/g, convertDirPathsToContextCalls(contexts));
+var output = source.replace(/COMPONENTS/g, convertPathMapToRequireCalls(components)).replace(/FIXTURES/g, convertPathMapToRequireCalls(fixtures)).replace(/PROXIES/g, convertPathListToRequireCalls(proxies)).replace(/CONTEXTS/g, convertDirPathsToContextCalls(contexts));
+console.log(output);
+return output;
```

Let me know what does webpack log when you run Cosmos afterwards.

PS. Also let me know if you have any idea on how to better collaborate on this.

**Comment by reinei** [comment]
The output is
```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
// The following constants are replaced at compile-time (through module-loader)
// with a map of require calls with absolute paths, derived from user conf.
// By the time webpack analyzes this file to build the dependency tree
// all paths will be embedded and webpack will register watchers for each
// context (which will update the bundle automatically when files are added or
// changed).

// eslint-disable-next-line no-undef, no-unused-expressions
[require.context('D:/Development/centrum-sys/src/components/Localisation/LocalisedText', false),require.context('D:/Development/centrum-sys/src/components/Localisation/LocalisedText/__fixtures__', false)];

var getUserModules = function getUserModules() {
  return {
    /* eslint-disable no-undef */
    components: {'Localisation/LocalisedText':require('D:/Development/centrum-sys/src/components/Localisation/LocalisedText/LocalisedText.js')},
    fixtures: {'Localisation/LocalisedText':{'LocalisedText':require('D:/Development/centrum-sys/src/components/Localisation/LocalisedText/__fixtures__/LocalisedText.js')}},
    proxies: [require('D:\Development\centrum-sys\redux-proxy.js'),require('D:\Development\centrum-sys\context-proxy.js')]
  };
};
```

Which has the wrong kind of slashes for the proxies (but the right kind for the components)

**Comment by skidding** [comment]
I see.

Well the component/fixture paths are fine because they're generated using node-glob, which [seems to work exclusively with forward slashes](https://github.com/isaacs/node-glob#windows). It seems like the general solution is to do this with *all* paths, because Node on Windows [supports it, even though it defaults to backward slashes](https://nodejs.org/api/path.html#path_path_win32).

[slash](https://github.com/sindresorhus/slash) looks like the right tool for the job. It should probably be applied [here](https://github.com/react-cosmos/react-cosmos/blob/ed8ef3e09645cd69d0086339ea6e169294bd4aef/packages/react-cosmos-config/src/index.js#L6-L8). Before we do that, though, you could try this:

```bash
npm i -S slash
```

```js
// cosmos.config.js
var slash = require('slash');
// ...
proxies: [
  slash(require.resolve('./context-proxy.js')),
  slash(require.resolve('./redux-proxy.js')),
]
```

**Comment by reinei** [comment]
Using slash as suggested above does indeed fix my issue.
Thanks for the help skidding (awesome project by the way)

**Comment by skidding** [comment]
Reopening because this should be fixed for all Windows users.

**Comment by RadValentin** [comment]
I think we can close this issue now, it was addressed in #303 cc: @skidding @reinei
